File: src/main/java/judoku/Solver.java
/*	Copyright (C) 2018 Steve Ball <jetpants@gmail.com>

	This file is part of Judoku. Judoku is free software: you can redistribute
	it and/or modify it under the terms of the GNU General Public License as
	published by the Free Software Foundation, either version 3 of the License,
	or (at your option) any later version.

    Judoku is distributed in the hope that it will be useful, but WITHOUT ANY
    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
    FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
    details.

    You should have received a copy of the GNU General Public License along with
    Judoku. If not, see <http://www.gnu.org/licenses/>.
*/

package judoku;

import java.util.AbstractList;
import java.util.ArrayList;

public class Solver {
    public Solver(Grid puzzle) {
        if (!puzzle.isLegal())
            throw new IllegalArgumentException("Grid has invalid state and no solutions");

        this.puzzle = new Grid(puzzle);
    }

    public Grid[] findSolutions(int maxSolutions) {
		AbstractList<Grid> solutions = new ArrayList<Grid>(maxSolutions);
		int n = search(new Grid(puzzle), solutions, maxSolutions);
		assert n == solutions.size() : n;
		Grid[] results = new Grid[n];
		return solutions.toArray(results);
	}

    public int countSolutions(int maxSolutions) {
		return search(new Grid(puzzle), null, maxSolutions);
	}

    public boolean isProper() {
        /*  A so-called 'proper' puzzle has exactly one solution and no given is superfluous.
            I.e., removing any one of the givens would result in a puzzle with more than one
            solution. Proper puzzles are both sufficent and minimal. */

        return hasUniqueSolution() && isMinimal();
    }

    public boolean hasUniqueSolution() {
        return countSolutions(2) == 1;          // one and only one
    }

    public boolean isMinimal() {
        boolean minimal = true;

        for (int nth = 1; nth <= puzzle.numCells() && minimal; ++nth) {
            int val = puzzle.getCell(nth);

            if (val != Grid.EMPTY) {
                puzzle.setEmpty(nth);
                minimal = countSolutions(2) > 1;
                puzzle.setCell(nth, val);
            }
        }

        return minimal;
    }

    // static variants ------------------------------------------------------------------------

    public static Grid[] findSolutions(Grid puzzle, int maxSolutions) {
        return new Solver(puzzle).findSolutions(maxSolutions);
    }

    public static int countSolutions(Grid puzzle, int maxSolutions) {
        return new Solver(puzzle).countSolutions(maxSolutions);
	}

    public static boolean isProper(Grid grid) {
        return new Solver(grid).isProper();
    }

    public static boolean hasUniqueSolution(Grid grid) {
        return new Solver(grid).hasUniqueSolution();
    }

    public static boolean isMinimal(Grid grid) {
        return new Solver(grid).isMinimal();
    }

    // ----------------------------------------------------------------------------------------

	private static int search(Grid g, AbstractList<Grid> solutions, int maxSolutions) {
		assert g.isLegal();
        // lots of strategies that could be added here: http://www.sudokudragon.com/sudokustrategy.htm
		return bruteForce(g, solutions, maxSolutions);
	}

	private static int bruteForce(Grid template, AbstractList<Grid> solutions, int max) {
        if (max == 0) return 0;

		boolean incomplete = false;			// an empty cell has been found
		int col = 1, row = 1;
		search:
			for (row = 1; row <= template.numRows(); ++row)
				for (col = 1; col <= template.numColumns(); ++col)
					if (template.getCell(col, row) == Grid.EMPTY) {
						incomplete = true;
						break search;
					}

		if (!incomplete) {
			// no more empty cells - a complete solution has been found :)
			if (solutions != null) {
				assert template.isLegal();
				solutions.add(new Grid(template));
			}
			return 1;
		}

		int[] values = template.getCandidates(col, row);
        Util.shuffle(values);

        int found = 0;

		for (int value: values) {
			template.setCell(col, row, value);
			found += bruteForce(template, solutions, max - found);
		}
		template.setCell(col, row, Grid.EMPTY);

		return found;
	}

    private final Grid puzzle;
}

File: src/main/kotlin/judoku/Solver.kt
/*	Copyright (C) 2018 Steve Ball <jetpants@gmail.com>

	This file is part of Judoku. Judoku is free software: you can redistribute
	it and/or modify it under the terms of the GNU General Public License as
	published by the Free Software Foundation, either version 3 of the License,
	or (at your option) any later version.

    Judoku is distributed in the hope that it will be useful, but WITHOUT ANY
    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
    FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
    details.

    You should have received a copy of the GNU General Public License along with
    Judoku. If not, see <http://www.gnu.org/licenses/>.
*/

package judoku

class Solver(private val puzzle: Grid) {
    init {
        check(puzzle.isLegal())
    }

    fun findSolutions(maxSolutions: Int): Array<Grid> {
        val solutions = ArrayList<Grid>(maxSolutions)
        val n = traverse(puzzle, solutions, maxSolutions)
        assert(n == solutions.size) { n }
        val results = arrayOfNulls<Grid>(n)
        return solutions.toArray(results)
    }

    fun countSolutions(maxSolutions: Int): Int = traverse(puzzle, null, maxSolutions)

    fun suggestMove(): Move? {
        /*  If the puzzle has duplicates or unusable cells, then no sequence of moves will
            yield a solution. */
        if (!puzzle.isViable()) return null;

        check(puzzle.isLegal());
        check(puzzle.isViable());

        val matrix = Matrix(puzzle)
        return matrix.suggestMove()
    }

    fun hasUniqueSolution(): Boolean  = countSolutions(2) == 1      // one and only one

    fun isMinimal(): Boolean {
        if (!puzzle.isViable()) return false;

        for (nth in 1..puzzle.numCells())
            if (!puzzle.isEmpty(nth) && countSolutions(puzzle.withCellEmpty(nth), 2) < 2)
                return false

        return true
    }

    fun isProper(): Boolean {
        /*  A so-called 'proper' puzzle has exactly one solution and no given is superfluous.
            I.e., removing any one of the givens would result in a puzzle with more than one
            solution. Proper puzzles are both sufficent and minimal. */
        return hasUniqueSolution() && isMinimal()
    }

    companion object {
        @JvmStatic
        fun findSolutions(puzzle: Grid, maxSolutions: Int): Array<Grid> =
            Solver(puzzle).findSolutions(maxSolutions)

        @JvmStatic
        fun countSolutions(puzzle: Grid, maxSolutions: Int): Int =
            Solver(puzzle).countSolutions(maxSolutions)

        @JvmStatic
        fun suggestMove(puzzle: Grid): Move? = Solver(puzzle).suggestMove()

        @JvmStatic
        fun hasUniqueSolution(grid: Grid): Boolean = Solver(grid).hasUniqueSolution()

        @JvmStatic
        fun isMinimal(grid: Grid): Boolean = Solver(grid).isMinimal()

        @JvmStatic
        fun isProper(grid: Grid): Boolean = Solver(grid).isProper()

        private fun traverse(template: Grid, solutions: ArrayList<Grid>?, maxSolutions: Int): Int {
            if (!template.isViable()) return 0;           // there are no solutions
            return _traverse(template, solutions, maxSolutions)
        }

        private fun _traverse(template: Grid, solutions: ArrayList<Grid>?, max: Int): Int {
            if (max == 0) return 0

            var nth = nextCell(template)

            if (nth == null) {
                // no more empty cells, a solution has been found
                check(template.isLegal())
                if (solutions != null) solutions.add(template)
                return 1
            }

            val possibilities = template.getPossibilities(nth)
            Util.shuffle(possibilities)

            var found = 0

            for (possibility in possibilities) {
                val next = template.withCell(nth, possibility)
                found += _traverse(next, solutions, max - found)
            }

            return found
        }

        private fun nextCell(g: Grid): Int? {
            /*  The search performs a depth-first traversal of the tree of all possible grid
                cell-value combinations. From a given board there is a tree of grids that
                descends from this node that terminates in leaf nodes that are either complete
                solutions or in nodes where the grids are incomplete but have no further
                possible moves. This method determines the order in which that sub-tree is
                searched.

                If there are two empty cells and one could have any of six possible values and
                another could only have two, then it makes sense to search the sub-tree
                descending from the latter cell first because, a priori, if there is a solution
                then there's a one-in-two chance of finding it here, as opposed to a one-in-six
                chance of finding it in the sub-tree off the other empty cell.

                This method finds the empty cell that has the fewest possible values. This
                method used to search simply for the first empty cell. When I changed it to look
                for the empty cell with the fewest possibilities, the time to find a grid's
                solution dropped to a third! Thanks to Norvig for the optimisation!
                http://norvig.com/sudoku.html */

            var best : Int? = null
            var count = Int.MAX_VALUE

            for (nth in 1..g.numCells())
                if (g.isEmpty(nth)) {
                    val c = g.getPossibilities(nth).size
                    if (c < count) {
                        best = nth
                        count = c
                    }
                }

            return best
        }
    }
}
